File: mini-claw-code-py/src/mini_claw_code_py/tools/read.py
from __future__ import annotations

import asyncio
from pathlib import Path
from typing import Any

from ..types import ToolDefinition


class ReadTool:
    def __init__(self) -> None:
        self._definition = ToolDefinition.new(
            "read",
            "Read the contents of a file.",
        ).param("path", "string", "The file path to read", True)

    @property
    def definition(self) -> ToolDefinition:
        return self._definition

    async def call(self, args: Any) -> str:
        path = args.get("path") if isinstance(args, dict) else None
        if not isinstance(path, str):
            raise ValueError("missing 'path' argument")
        try:
            return await asyncio.to_thread(Path(path).read_text)
        except Exception as exc:
            raise RuntimeError(f"failed to read '{path}'") from exc

File: mini-claw-code-ts/src/tools/read.ts
import { readFile } from "node:fs/promises";
import { ToolDefinition, type JsonObject, type Tool } from "../types";

export class ReadTool implements Tool {
  readonly #definition = new ToolDefinition("read", "Read the contents of a file.").param(
    "path",
    "string",
    "The file path to read",
    true,
  );

  definition(): ToolDefinition {
    return this.#definition;
  }

  async call(args: JsonObject | null): Promise<string> {
    const path = args?.path;
    if (typeof path !== "string") {
      throw new Error("missing 'path' argument");
    }
    return readFile(path, "utf8");
  }
}
